### PR TITLE
add zk register code

### PR DESCRIPTION
**Pull request by pantianying**
**What this PR does**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: remoting/zookeeper/client.go [hunk 1]
```diff
@@ -35,14 +35,19 @@ import (
...
 	MaxFailTimes = 15
 )
 
 var (
 	errNilZkClientConn = perrors.New("zookeeperclient{conn} is nil")
+	errNilChildren     = perrors.Errorf("has none children")
+	errNilNode         = perrors.Errorf("node does not exist")
 )
 
+// ZookeeperClient ...
```
Comment (zouyx) [inline remoting/zookeeper/client.go]: add some comment for this struct

File: remoting/zookeeper/listener.go [hunk 2]
```diff
@@ -19,40 +19,43 @@ package zookeeper
...
 	perrors "github.com/pkg/errors"
-	"github.com/samuel/go-zookeeper/zk"
 )
 
 import (
 	"github.com/apache/dubbo-go/common/logger"
 	"github.com/apache/dubbo-go/remoting"
 )
 
+// ZkEventListener ...
```
Comment (zouyx) [inline remoting/zookeeper/listener.go]: add some comment for this struct

File: registry/base_register.go [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+
+	case common.PROVIDER:
+		dubboPath, rawURL, err = r.providerRegistry(c, params)
+	case common.CONSUMER:
+		dubboPath, rawURL, err = r.consumerRegistry(c, params)
+	default:
+		return perrors.Errorf("@c{%v} type is not referencer or provider", c)
+	}
+	encodedURL = url.QueryEscape(rawURL)
+	dubboPath = strings.ReplaceAll(dubboPath, "$", "%24")
```
Comment (flycash) [inline registry/base_register.go]: I don't understand why not use url.QueryEscape for dubboPath?

File: registry/base_register.go [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+	}
+	dubboPath = fmt.Sprintf("/dubbo/%s/%s", r.service(c), common.DubboNodes[common.PROVIDER])
+	r.cltLock.Lock()
+	err = r.facadeBasedRegistry.CreatePath(dubboPath)
+	r.cltLock.Unlock()
+	if err != nil {
+		logger.Errorf("facadeBasedRegistry.CreatePath(path{%s}) = error{%#v}", dubboPath, perrors.WithStack(err))
+		return "", "", perrors.WithMessagef(err, "facadeBasedRegistry.CreatePath(path:%s)", dubboPath)
+	}
+	params.Add("anyhost", "true")
```
Comment (flycash) [inline registry/base_register.go]: What does it mean?

File: registry/base_register.go [hunk 5]
```diff
@@ -0,0 +1,375 @@
...
+	// Dubbo java consumer to start looking for the provider url,because the category does not match,
+	// the provider will not find, causing the consumer can not start, so we use consumers.
+	// DubboRole               = [...]string{"consumer", "", "", "provider"}
+	// params.Add("category", (RoleType(PROVIDER)).Role())
+	params.Add("category", (common.RoleType(common.PROVIDER)).String())
+	params.Add("dubbo", "dubbo-provider-golang-"+constant.Version)
+
+	params.Add("side", (common.RoleType(common.PROVIDER)).Role())
+
+	if len(c.Methods) == 0 {
```
Comment (flycash) [inline registry/base_register.go]: `if len(c.Methods) != 0 { ...}`????

File: remoting/etcdv3/client.go [hunk 6]
```diff
@@ -36,16 +36,22 @@ import (
...
 
 var (
+	// ErrNilETCDV3Client ...
 	ErrNilETCDV3Client = perrors.New("etcd raw client is nil") // full describe the ERR
-	ErrKVPairNotFound  = perrors.New("k/v pair not found")
+	// ErrKVPairNotFound ...
+	ErrKVPairNotFound = perrors.New("k/v pair not found")
 )
 
+// Options ...
```
Comment (zouyx) [inline remoting/etcdv3/client.go]: add some comment for this struct

File: registry/base_register.go [hunk 7]
```diff
@@ -0,0 +1,375 @@
...
+	r.cltLock.Lock()
+	err = r.facadeBasedRegistry.CreatePath(dubboPath)
+	r.cltLock.Unlock()
+	if err != nil {
+		logger.Errorf("facadeBasedRegistry.CreatePath(path{%s}) = error{%v}", dubboPath, perrors.WithStack(err))
+		return "", "", perrors.WithStack(err)
+	}
+	dubboPath = fmt.Sprintf("/dubbo/%s/%s", r.service(c), common.DubboNodes[common.PROVIDER])
+	r.cltLock.Lock()
+	err = r.facadeBasedRegistry.CreatePath(dubboPath)
```
Comment (flycash) [inline registry/base_register.go]: Why should we create provider's path here?

File: registry/zookeeper/listener.go [hunk 8]
```diff
@@ -36,18 +36,23 @@ import (
 	zk "github.com/apache/dubbo-go/remoting/zookeeper"
 )
 
+// RegistryDataListener ...
```
Comment (zouyx) [inline registry/zookeeper/listener.go]: add some comment for this file

File: registry/base_register.go [hunk 9]
```diff
@@ -0,0 +1,375 @@
...
+	r.cltLock.Unlock()
+	if err != nil {
+		logger.Errorf("facadeBasedRegistry.CreatePath(path{%s}) = error{%v}", dubboPath, perrors.WithStack(err))
+		return "", "", perrors.WithStack(err)
+	}
+
+	params.Add("protocol", c.Protocol)
+	params.Add("category", (common.RoleType(common.CONSUMER)).String())
+	params.Add("dubbo", "dubbogo-consumer-"+constant.Version)
```
Comment (flycash) [inline registry/base_register.go]: `params.Add("dubbo", "dubbo-provider-golang-"+constant.Version)` but here is lack of `golang`. Is that right??

File: registry/base_register.go [hunk 10]
```diff
@@ -0,0 +1,375 @@
...
+	rawURL = fmt.Sprintf("consumer://%s%s?%s", localIP, c.Path, params.Encode())
+	dubboPath = fmt.Sprintf("/dubbo/%s/%s", r.service(c), (common.RoleType(common.CONSUMER)).String())
+
+	logger.Debugf("consumer path:%s, url:%s", dubboPath, rawURL)
+	return dubboPath, rawURL, nil
+}
+
+// sleepWait...
+func sleepWait(n int) {
+	wait := time.Duration((n + 1) * 2e8)
```
Comment (flycash) [inline registry/base_register.go]: `2e8` is hard to understand, please define a constant.

File: registry/base_register.go [hunk 11]
```diff
@@ -0,0 +1,375 @@
...
+		}
+
+		listener, err := r.facadeBasedRegistry.DoSubscribe(url)
+		if err != nil {
+			if !r.IsAvailable() {
+				logger.Warnf("event listener game over.")
+				return
+			}
+			logger.Warnf("getListener() = err:%v", perrors.WithStack(err))
+			time.Sleep(time.Duration(RegistryConnDelay) * time.Second)
```
Comment (flycash) [inline registry/base_register.go]: Why not just define the RegistryConnDelay as `time.Duration(3* time.Second)`

File: remoting/zookeeper/listener.go [hunk 12]
```diff
@@ -263,56 +280,11 @@ func timeSecondDuration(sec int) time.Duration {
 	return time.Duration(sec) * time.Second
 }
 
-// this func is invoked by ZkConsumerRegistry::Register/ZkConsumerRegistry::get/ZkConsumerRegistry::getListener
+// ListenServiceEvent is invoked by ZkConsumerRegistry::Register/ZkConsumerRegistry::get/ZkConsumerRegistry::getListener
 // registry.go:Listen -> listenServiceEvent -> listenDirEvent -> ListenServiceNodeEvent
 //                            |
 //                            --------> ListenServiceNodeEvent
 func (l *ZkEventListener) ListenServiceEvent(zkPath string, listener remoting.DataListener) {
-	var (
```
Comment (flycash) [inline remoting/zookeeper/listener.go]: Do You remove this part? I don't find the zk's DataListener implementation contains this logics.

**Pull request merged**